example/grammar/magnet.cpp

```diff
@@ -0,0 +1,496 @@
+    param(string_view key) const noexcept
+    {
+        params_encoded_view ps = u_.encoded_params();
+        auto it = ps.begin();
+        auto end = ps.end();
+        if (it != end)
+        {
+            query_param_encoded_view p = *it;
+            if (p.key.starts_with("x.") &&
+                p.key.substr(2) == key &&
+                p.has_value)
+                return pct_encoded_view(p.value);
+            ++it;
+        }
+        return boost::none;
+    }
+
+    friend
+    std::ostream&
+    operator<<(std::ostream& os, magnet_link_view m)
+    {
+        return os << m.u_;
+    }
+
+private:
+    // get a parameter as a pct_encoded_view
+    optional<pct_encoded_view>
+    decode_param(string_view key) const noexcept
+    {
+        params_encoded_view ps = u_.encoded_params();
+        auto it = ps.find(key);
+        if (it != ps.end() && (*it).has_value)
+            return pct_encoded_view((*it).value);
+        return boost::none;
+    }
+
+    // get a parameter as a url_view
+    optional<url_view>
+    url_param(string_view key) const noexcept
+    {
+        params_encoded_view ps = u_.encoded_params();
+        auto it = ps.find(key);
+        if (it != ps.end() && (*it).has_value)
+        {
+            result<url_view> r = parse_uri((*it).value);
+            if (r)
+                return *r;
+        }
+        return boost::none;
+    }
+
+    friend magnet_link_rule_t;
+};
+
+struct magnet_link_rule_t
+{
+    using value_type = magnet_link_view;
+
+    // Specs:
+    // https://www.bittorrent.org/beps/bep_0005.html
+    // https://www.bittorrent.org/beps/bep_0009.html
+    // https://www.bittorrent.org/beps/bep_0053.html
+    // https://github.com/webtorrent/magnet-uri
+    // https://en.wikipedia.org/wiki/Magnet_URI_scheme
+
+    result< value_type >
+    parse( char const*& it, char const* end ) const noexcept
+    {
+        // 1) Check if string starts with url
+        // (absolute_uri_rule_t should return url_view)
+        char const* it0 = it;
+        result<absolute_uri_rule_t::value_type> rrv =
+            grammar::parse(it, end, absolute_uri_rule);
+        if(!rrv)
+        {
+            it = it0;
+            return error::invalid;
+        }
+
+        // 2) Parse this url
+        magnet_link_view m;
+        m.u_ = *parse_absolute_uri({it0, it});
+
+        // 3) Check if exact topics are valid urls
+        // and that we have at least one
+        auto ps = m.u_.encoded_params();
+        auto pit = ps.begin();
+        auto pend = ps.end();
+        auto f = magnet_link_view::is_exact_topic{};
+        pit = std::find_if(pit, pend, f);
+        if (pit == pend)
+        {
+            // no topic in the magnet link
+            return error::invalid;
+        }
+
+        // all topics parse as valid urls
+        if (!std::all_of(pit, pend, [](
+            query_param_encoded_view p)
+        {
+            if (!magnet_link_view::is_exact_topic{}(p))
+                return true;
```

Comment:
wait how can you return true, shouldn't this be an error code?

Reply:
Nop... We are checking if all exact topics are valid urls. All other params are valid.